src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2109 @@
         B18 A8
         B19 A19
         B20 A20
-        B21 N/C
+        B21 WE (for Flashrom)
         B22 A21
         B23 VCC
         B24 N/C
         B25 GND
 
+    Storyware
+    ---------
+
+    Although no booklet artwork is rendered, toggling input port 'Pen Target'
+    switches between mapping pen coordinates to the tablet or the Storyware.
+
+    Test Mode
+    ---------
+
+    Most games contain a hidden test mode that can be activated by the same inputs:
+
+        Pages 1, 3, and 5 covered, the others exposed;
+        Left red button held down;
+
+    If the machine configuration 'Test Mode Pages' is enabled, the driver forces this page setup,
+    so that the player only needs to hold the left red button at reset to activate this mode.
+
+    Toggling 'Memory Cache' allows the player to observe differences between test failure and success.
+
 *******************************************************************************/
 
 #include "emu.h"
 
 #include "bus/generic/slot.h"
 #include "bus/generic/carts.h"
-#include "cpu/arm7/arm7.h"
+#include "cpu/arm7/ap2010cpu.h"
+#include "machine/clock.h"
+#include "machine/timer.h"
+#include "sound/ap2010pcm.h"
 
+#include "crsshair.h"
+#include "emupal.h"
+#include "render.h"
 #include "softlist_dev.h"
+#include "schedule.h"
 #include "speaker.h"
 #include "screen.h"
 
+#include "beena.lh"
+
+#define VERBOSE (0)
+#include "logmacro.h"
 
 namespace {
 
-class sega_beena_state : public driver_device
+#define SCREEN_W 704
+#define SCREEN_H 480
+
+#define ROM_MASK_BASE 0x80000000
+#define ROM_FLASH_BASE 0xa0000000
+#define UNKNOWN_ADDR 0xffffffff
+
+#define MEMCACHE_FIFO_MAX_SIZE 0x100
```

Comment:
Please change these to `constexpr` variables – it makes for better debug symbols and less confusing errors.